chall06/vsempere.c

```diff
@@ -0,0 +1,218 @@
+{
+    int         row_chk;
+    int         max;
+
+    max = board_size * board_size;
+    row_chk = king - (ring *  board_size);
+    
+    return (row_chk >= 0 &&
+        (board[row_chk] & (VSEMPERE_Q | VSEMPERE_R)) ? 1 : 0);
+}
+
+static int      ft_check_south(char *board, int ring, int board_size, int king)
+{
+    int         row_chk;
+    int         max;
+
+    max = board_size * board_size;
+    row_chk = king + (ring *  board_size);
+    
+    return (row_chk < max &&
+        (board[row_chk] & (VSEMPERE_Q | VSEMPERE_R)) ? 1 : 0);
+}
+
+static int      ft_check_east(char *board, int ring, int board_size, int king)
+{
+    int         row_chk;
+    int         max;
+
+    max = board_size * board_size;
+    row_chk = king + ring;
+    
+    return (((king) % board_size < (row_chk) % board_size) && 
+        row_chk < max &&
+        (board[row_chk] & (VSEMPERE_Q | VSEMPERE_R)) ? 1 : 0);
+}
+
+static int      ft_check_west(char *board, int ring, int board_size, int king)
+{
+    int         row_chk;
+    int         max;
+
+    max = board_size * board_size;
+    row_chk = king - ring;
+    
+    return (((king) % board_size > (row_chk) % board_size) && 
+        row_chk >= 0 &&
+        (board[row_chk] & (VSEMPERE_Q | VSEMPERE_R)) ? 1 : 0);
+}
+
+static int     ft_check_mate_ring(char *board, int king,
+    int ring, int board_size)
+{
+    if (ring >= board_size)
+        return (2);
+
+    return (ft_check_south_east(board, ring, board_size, king) || 
+        ft_check_north_west(board, ring, board_size, king) ||
+        ft_check_south_west(board, ring, board_size, king) ||
+        ft_check_north_west(board, ring, board_size, king) ||
+        ft_check_north_east(board, ring, board_size, king) ||
+        ft_check_north(board, ring, board_size, king) ||
+        ft_check_south(board, ring, board_size, king) ||
+        ft_check_east(board, ring, board_size, king) ||
+        ft_check_west(board, ring, board_size, king) ? 1 : 0);
+}
+
+static int      ft_check_map(char *board)
+{
+    int pos = 0;
+    int board_side = 0;
+    int king = 0;
+
+    if (!board)
+        return (0);
+    while (*board)
+    {
+        if (!board_side && *board=='\n')
+        {
+            board++;
+            board_side = pos;
+            continue ;
+        }
+        else if (board_side && *board=='\n')
+        {
+            if (pos % board_side)
+                return (0);
+            board++;
+            continue;
+        }
+        else if (!strchr("PBRQK.\n", *board))
+            return (0);
+        else if (*board=='K')
+            king++;
+        board++;
+        pos++;
+    }
+    return (king != 1 ? 0 : 1);
+}
+
+int             ft_check_mate(char *board_param)
+{
+    int hit;
+    int king;
+    int board_side = 0;
+    int ring;
+    char *board;
+
+    if (!ft_check_map(board_param))
+        return (1);
+    if (!(board = malloc(sizeof(char) * strlen(board_param))))
+        return (1);
+    strcpy(board, board_param);
```

Comment:
`strncpy` > `strcpy`. Your `strcpy` call heap-overflow's because you miscalculated the size for the malloc :( this is very dangerous, you are writing in random memory.

This leads to all sort of undefined behaviour. Because of this invalid write, you probably fail the checks for `board = "K.....\n......\n......\n......"`. You return `0`. This board is obviously invalid, and also the King is safe, so it's a double error.